README.md
check_hp_firmware
=================

![Go build](https://github.com/NETWAYS/check_hp_firmware/workflows/Go/badge.svg?branch=master)

<!-- Note: Update `Readme` in main.go when changing this! -->

Icinga / Nagios check plugin to verify HPE controllers an SSD disks are not affected by certain vulnerabilities.

**HPE Controllers**

> HPE Smart Array SR Gen10 Controller Firmware Version 2.65 (or later) provided in the (HPE document a00097210) is
> required to prevent a potential data inconsistency on select RAID configurations with Smart Array Gen10 Firmware
> Version 1.98 through 2.62, based on the following scenarios. HPE strongly recommends performing this upgrade at the
> customer's earliest opportunity per the "Action Required" in the table located in the Resolution section.
> Neglecting to perform the recommended resolution could result in potential subsequent errors and potential data
> inconsistency.

The check will alert you with a CRITICAL when the firmware is in the affected range with:

* `if you have RAID 1/10/ADM - update immediately!`
* `if you have RAID 5/6/50/60 - update immediately!`

And it will add a short note when `firmware older than affected` or `firmware has been updated`. At the moment the
plugin does not verify configured logical drives, but we believe you should update in any case.

**HPE SSD SAS disks**

> HPE SAS Solid State Drives - Critical Firmware Upgrade Required for Certain HPE SAS Solid State Drive Models to
> Prevent Drive Failure at 32,768 or 40,000 Hours of Operation

The check will raise a CRITICAL when the drive needs to be updated with the note `affected by FW bug`, and when
the drive is patched with `firmware update applied`.

Please see support documents from HPE:
* [a00092491](https://support.hpe.com/hpesc/public/docDisplay?docLocale=en_US&docId=emr_na-a00092491en_us)
* [a00097382](https://support.hpe.com/hpesc/public/docDisplay?docLocale=en_US&docId=a00097382en_us)
* [a00097210](https://support.hpe.com/hpesc/public/docDisplay?docLocale=en_US&docId=a00097210en_us)

**IMPORTANT:** Read the documentation for HPE! The plugin and its documentation is a best effort to find and detect
affected hardware. There is ABSOLUTELY NO WARRANTY, see the license!

> **Note:** This plugin was initially named check_hp_disk_firmware

## Usage

Arguments:

      -H, --hostname string        SNMP host (default "localhost")
      -c, --community string       SNMP community (default "public")
      -P, --protocol string        SNMP protocol (default "2c")
          --timeout int            SNMP timeout in seconds (default 15)
          --snmpwalk-file string   Read output from snmpwalk
      -4, --ipv4                   Use IPv4
      -6, --ipv6                   Use IPv6
      -V, --version                Show version
          --debug                  Enable debug output

Simply run the command:

    $ ./check_hp_firmware -H localhost -c public

## Installation

This is a golang project, either download the binary from the releases:

https://github.com/NETWAYS/check_hp_firmware/releases

Also see the included [CheckCommand for Icinga 2](icinga2.conf).

You can download or build the project locally with go:

    $ go get github.com/NETWAYS/check_hp_cve
    
    $ git clone https://github.com/NETWAYS/check_hp_firmware
    $ cd check_hp_firmware/
    $ go build -o check_hp_firmware .

## Example

    OK - All 2 controllers and 33 drives seem fine
    [OK] controller (0) model=p816i-a serial=XXX firmware=1.65 - firmware older than affected
    [OK] controller (4) model=p408e-p serial=XXX firmware=1.65 - firmware older than affected
    [OK] (0.9 ) model=MO003200JWFWR serial=XXX firmware=HPD2 hours=8086
    [OK] (0.11) model=EK000400GWEPE serial=XXX firmware=HPG0 hours=8086
    [OK] (0.12) model=EK000400GWEPE serial=XXX firmware=HPG0 hours=8086
    [OK] (0.14) model=MO003200JWFWR serial=XXX firmware=HPD2 hours=8086
    [OK] (4.0 ) model=MO3200JFFCL serial=XXX firmware=HPD8 hours=7568 - firmware update applied
    [OK] (4.1 ) model=MO3200JFFCL serial=XXX firmware=HPD8 hours=7568 - firmware update applied
    [OK] (4.2 ) model=MO3200JFFCL serial=XXX firmware=HPD8 hours=7568 - firmware update applied
    [OK] (4.3 ) model=MO3200JFFCL serial=XXX firmware=HPD8 hours=7568 - firmware update applied
    [OK] (4.4 ) model=MO3200JFFCL serial=XXX firmware=HPD8 hours=7568 - firmware update applied
    [OK] (4.5 ) model=MO3200JFFCL serial=XXX firmware=HPD8 hours=7568 - firmware update applied
    [OK] (4.6 ) model=MO3200JFFCL serial=XXX firmware=HPD8 hours=7568 - firmware update applied
    [OK] (4.24) model=MO3200JFFCL serial=XXX firmware=HPD8 hours=7568 - firmware update applied
    [OK] (4.25) model=MO3200JFFCL serial=XXX firmware=HPD8 hours=7568 - firmware update applied
    [OK] (4.26) model=MO3200JFFCL serial=XXX firmware=HPD8 hours=7568 - firmware update applied
    [OK] (4.27) model=MO3200JFFCL serial=XXX firmware=HPD8 hours=7568 - firmware update applied
    [OK] (4.28) model=MO3200JFFCL serial=XXX firmware=HPD8 hours=7568 - firmware update applied
    [OK] (4.29) model=MO3200JFFCL serial=XXX firmware=HPD8 hours=7568 - firmware update applied
    [OK] (4.30) model=MO3200JFFCL serial=XXX firmware=HPD8 hours=7568 - firmware update applied
    [OK] (4.31) model=MO3200JFFCL serial=XXX firmware=HPD8 hours=7568 - firmware update applied
    [OK] (4.50) model=MO3200JFFCL serial=XXX firmware=HPD8 hours=7568 - firmware update applied
    [OK] (4.51) model=MO003200JWFWR serial=XXX firmware=HPD2 hours=7568
    [OK] (4.52) model=MO3200JFFCL serial=XXX firmware=HPD8 hours=7568 - firmware update applied
    [OK] (4.53) model=MO3200JFFCL serial=XXX firmware=HPD8 hours=7568 - firmware update applied
    [OK] (4.54) model=MO3200JFFCL serial=XXX firmware=HPD8 hours=7568 - firmware update applied
    [OK] (4.55) model=MO3200JFFCL serial=XXX firmware=HPD8 hours=7568 - firmware update applied
    [OK] (4.56) model=MO3200JFFCL serial=XXX firmware=HPD8 hours=7568 - firmware update applied
    [OK] (4.75) model=MO3200JFFCL serial=XXX firmware=HPD8 hours=7568 - firmware update applied
    [OK] (4.76) model=MO3200JFFCL serial=XXX firmware=HPD8 hours=7568 - firmware update applied
    [OK] (4.77) model=MO3200JFFCL serial=XXX firmware=HPD8 hours=7568 - firmware update applied
    [OK] (4.78) model=MO3200JFFCL serial=XXX firmware=HPD8 hours=7568 - firmware update applied
    [OK] (4.79) model=MO3200JFFCL serial=XXX firmware=HPD8 hours=7568 - firmware update applied
    [OK] (4.80) model=MO3200JFFCL serial=XXX firmware=HPD8 hours=7568 - firmware update applied
    [OK] (4.81) model=MO3200JFFCL serial=XXX firmware=HPD8 hours=7568 - firmware update applied

## Limitations

* No SNMPv3 support is implemented

## Contribute

If you find bugs or want to add features, please open an issue or pull-request on GitHub.

You can help with problems by supplying the output of snmpwalk for the system you experience problems with:

    $ snmpwalk -c public -v2c -On HOST 1.3.6.1.4.1.232
    
Please make sure to either censor the output of any private information, or send an e-mail to support@netways.de,
so we can provide you with a secure upload link, that won't be shared with public.

## Technical Details

Supported hardware is split into modules: [hp/cntlr](hp/cntlr) [hp/phy_drv](hp/phy_drv)

Known models and affected firmware is documented in: [hp/cntlr/firmware_data.go](hp/cntlr/firmware_data.go) [hp/phy_drv/firmware_data.go](hp/phy_drv/firmware_data.go)

This data can be easily enhanced in the future. Make sure to document source documents and versions as well, and check
the accompanying firmware and status functions.

The check reads the `cpqDaCntlrTable` and `cpqDaPhyDrvTable` tables from SNMP, which should be available over the
IPMI agent or the locally installed HP tools, hooked into the SNMP daemon of the operating system.

## License

Copyright (C) 2020 Markus Frosch <markus.frosch@netways.de>

Copyright (C) 2020 NETWAYS <info@netways.de>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.


CHANGELOG.md
# Changelog

## [v1.2.0](https://github.com/NETWAYS/check_hp_firmware/tree/v1.2.0) (2020-04-09)

[Full Changelog](https://github.com/NETWAYS/check_hp_firmware/compare/v1.1.0...v1.2.0)

**Implemented enhancements:**

- Rename plugin to check\_hp\_firmware [\#10](https://github.com/NETWAYS/check_hp_firmware/issues/10)
- Rename plugin to check\_hp\_firmware [\#11](https://github.com/NETWAYS/check_hp_firmware/pull/11) ([lazyfrosch](https://github.com/lazyfrosch))
- Add support for controllers and bulletin a00097210 [\#9](https://github.com/NETWAYS/check_hp_firmware/pull/9) ([lazyfrosch](https://github.com/lazyfrosch))

## [v1.1.0](https://github.com/NETWAYS/check_hp_firmware/tree/v1.1.0) (2020-03-27)

[Full Changelog](https://github.com/NETWAYS/check_hp_firmware/compare/v1.0.1...v1.1.0)

**Implemented enhancements:**

- Add support for the a00097382 bulletin [\#4](https://github.com/NETWAYS/check_hp_firmware/pull/4) ([lazyfrosch](https://github.com/lazyfrosch))

**Merged pull requests:**

- Bump github.com/sirupsen/logrus from 1.4.2 to 1.5.0 [\#8](https://github.com/NETWAYS/check_hp_firmware/pull/8) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- Bump github.com/spf13/pflag from 1.0.3 to 1.0.5 [\#7](https://github.com/NETWAYS/check_hp_firmware/pull/7) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- Bump github.com/soniah/gosnmp from 1.22.0 to 1.24.0 [\#6](https://github.com/NETWAYS/check_hp_firmware/pull/6) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))
- Bump github.com/stretchr/testify from 1.3.0 to 1.5.1 [\#5](https://github.com/NETWAYS/check_hp_firmware/pull/5) ([dependabot-preview[bot]](https://github.com/apps/dependabot-preview))

## [v1.0.1](https://github.com/NETWAYS/check_hp_firmware/tree/v1.0.1) (2020-01-30)

[Full Changelog](https://github.com/NETWAYS/check_hp_firmware/compare/v1.0.0...v1.0.1)

**Fixed bugs:**

- Ensure panics are correctly handled with exit 3 [\#2](https://github.com/NETWAYS/check_hp_firmware/issues/2)
- Catch panics with nagios helpers to exit 3 [\#3](https://github.com/NETWAYS/check_hp_firmware/pull/3) ([lazyfrosch](https://github.com/lazyfrosch))
- Fix snmp values to be int/uint 32-bit values [\#1](https://github.com/NETWAYS/check_hp_firmware/pull/1) ([lazyfrosch](https://github.com/lazyfrosch))

## [v1.0.0](https://github.com/NETWAYS/check_hp_firmware/tree/v1.0.0) (2020-01-26)

[Full Changelog](https://github.com/NETWAYS/check_hp_firmware/compare/7c8517e36fc61fb7641fa90a6ab2b009765a1034...v1.0.0)



\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


main.go
package main

import (
	"fmt"
	"github.com/NETWAYS/check_hp_firmware/hp/cntlr"
	"github.com/NETWAYS/check_hp_firmware/hp/phy_drv"
	"github.com/NETWAYS/check_hp_firmware/nagios"
	"github.com/NETWAYS/check_hp_firmware/snmp"
	log "github.com/sirupsen/logrus"
	"github.com/soniah/gosnmp"
	flag "github.com/spf13/pflag"
	"io"
	"os"
	"time"
)

const Readme = `
Icinga / Nagios check plugin to verify HPE controllers an SSD disks are not affected by certain vulnerabilities.

**HPE Controllers**

  HPE Smart Array SR Gen10 Controller Firmware Version 2.65 (or later) provided in the (HPE document a00097210) is
  required to prevent a potential data inconsistency on select RAID configurations with Smart Array Gen10 Firmware
  Version 1.98 through 2.62, based on the following scenarios. HPE strongly recommends performing this upgrade at the
  customer's earliest opportunity per the "Action Required" in the table located in the Resolution section.
  Neglecting to perform the recommended resolution could result in potential subsequent errors and potential data
  inconsistency.

The check will alert you with a CRITICAL when the firmware is in the affected range with:

* "if you have RAID 1/10/ADM - update immediately!"
* "if you have RAID 5/6/50/60 - update immediately!"

And it will add a short note when "firmware older than affected" or "firmware has been updated". At the moment the
plugin does not verify configured logical drives, but we believe you should update in any case.

**HPE SSD SAS disks**

  HPE SAS Solid State Drives - Critical Firmware Upgrade Required for Certain HPE SAS Solid State Drive Models to
  Prevent Drive Failure at 32,768 or 40,000 Hours of Operation

The check will raise a CRITICAL when the drive needs to be updated with the note "affected by FW bug", and when
the drive is patched with "firmware update applied".

Please see support documents from HPE:
* https://support.hpe.com/hpesc/public/docDisplay?docLocale=en_US&docId=emr_na-a00092491en_us
* https://support.hpe.com/hpesc/public/docDisplay?docLocale=en_US&docId=a00097382en_us
* https://support.hpe.com/hpesc/public/docDisplay?docLocale=en_US&docId=a00097210en_us

**IMPORTANT:** Read the documentation for HPE! The plugin and its documentation is a best effort to find and detect
affected hardware. There is ABSOLUTELY NO WARRANTY, see the license!

**Note:** This plugin was initially named check_hp_disk_firmware
`

// Check for HP Controller CVEs via SNMP
func main() {
	flagSet := flag.NewFlagSet(os.Args[0], flag.ContinueOnError)
	flagSet.SortFlags = false

	host := flagSet.StringP("hostname", "H", "localhost", "SNMP host")
	community := flagSet.StringP("community", "c", "public", "SNMP community")
	protocol := flagSet.StringP("protocol", "P", "2c", "SNMP protocol")
	timeout := flagSet.Int64("timeout", 15, "SNMP timeout in seconds")

	file := flagSet.String("snmpwalk-file", "", "Read output from snmpwalk")

	ipv4 := flagSet.BoolP("ipv4", "4", false, "Use IPv4")
	ipv6 := flagSet.BoolP("ipv6", "6", false, "Use IPv6")

	version := flagSet.BoolP("version", "V", false, "Show version")

	debug := flagSet.Bool("debug", false, "Enable debug output")

	flagSet.Usage = func() {
		fmt.Printf("Usage: %s [-H <hostname>] [-c <community>]\n", os.Args[0])
		fmt.Println(Readme)
		fmt.Printf("Version: %s\n", buildVersion())
		fmt.Println()
		fmt.Println("Arguments:")
		flagSet.PrintDefaults()
	}

	var err error
	err = flagSet.Parse(os.Args[1:])
	if err != nil {
		if err != flag.ErrHelp {
			nagios.ExitError(err)
		}
		os.Exit(3)
	}

	if *version {
		fmt.Printf("%s version %s\n", Project, buildVersion())
		os.Exit(nagios.Unknown)
	}

	if *debug {
		log.SetLevel(log.DebugLevel)
	} else {
		defer nagios.CatchPanic()
	}

	var client *gosnmp.GoSNMP
	var cntlrTable *cntlr.CpqDaCntlrTable
	var driveTable *phy_drv.CpqDaPhyDrvTable

	if *file != "" {
		var fh *os.File
		fh, err = os.Open(*file)
		if err != nil {
			nagios.ExitError(err)
		}
		defer fh.Close()

		cntlrTable, err = cntlr.LoadCpqDaCntlrTable(fh)
		if err != nil {
			nagios.ExitError(err)
		}

		// jump back to start
		_, err = fh.Seek(0, io.SeekStart)
		if err != nil {
			nagios.ExitError(err)
		}

		driveTable, err = phy_drv.LoadCpqDaPhyDrvTable(fh)
		if err != nil {
			nagios.ExitError(err)
		}
	} else {
		defaultClient := *gosnmp.Default
		client = &defaultClient
		client.Target = *host
		client.Community = *community
		client.Timeout = time.Duration(*timeout) * time.Second
		client.Retries = 1

		if err := snmp.SetVersion(client, *protocol); err != nil {
			nagios.ExitError(err)
		}

		if *ipv4 {
			err = client.ConnectIPv4()
		} else if *ipv6 {
			err = client.ConnectIPv6()
		} else {
			err = client.Connect()
		}
		if err != nil {
			nagios.ExitError(err)
		}
		defer client.Conn.Close()

		cntlrTable, err = cntlr.GetCpqDaCntlrTable(client)
		if err != nil {
			nagios.ExitError(err)
		}

		driveTable, err = phy_drv.GetCpqDaPhyDrvTable(client)
		if err != nil {
			nagios.ExitError(err)
		}
	}

	if len(cntlrTable.Snmp.Values) == 0 {
		nagios.Exit(3, "No HP controller data found!")
	}

	controllers, err := cntlr.GetControllersFromTable(cntlrTable)
	if err != nil {
		nagios.ExitError(err)
	}

	if len(driveTable.Snmp.Values) == 0 {
		nagios.Exit(3, "No HP drive data found!")
	}

	drives, err := phy_drv.GetPhysicalDrivesFromTable(driveTable)
	if err != nil {
		nagios.ExitError(err)
	}

	overall := nagios.Overall{}

	countControllers := 0
	for _, controller := range controllers {
		controllerStatus, desc := controller.GetNagiosStatus()
		overall.Add(controllerStatus, desc)
		countControllers += 1
	}

	countDrives := 0
	for _, drive := range drives {
		driveStatus, desc := drive.GetNagiosStatus()
		overall.Add(driveStatus, desc)
		countDrives += 1
	}

	status := overall.GetStatus()
	var summary string
	switch status {
	case nagios.OK:
		summary = fmt.Sprintf("All %d controllers and %d drives seem fine", countControllers, countDrives)
	case nagios.Warning:
		summary = fmt.Sprintf("Found %d warnings", overall.Warnings)
	case nagios.Critical:
		summary = fmt.Sprintf("Found %d critical problems", overall.Criticals)
	}
	overall.Summary = summary
	nagios.Exit(status, overall.GetOutput())
}
